zendesk/importer.go

```diff
@@ -0,0 +1,187 @@
+	From interface{} `json:"from"`
+	To   interface{} `json:"to"`
+	Rel  interface{} `json:"rel"`
+}
+
+type BulkImporter struct {
+	userName   string       // userName to login as admin to zendesk
+	apiToken   string       // token to authenticate the user
+	client     *http.Client // http client to connect zendesk
+	maxRetries uint64       // max API retries in case of 429, before returning error
+	baseURL    string       // zendesk api url
+	retryCount uint64       // number of retry count made for current data
+}
+
+// NewBulkImporter initialize bulk importer to write bulk tickets to zendesk
+func NewBulkImporter(userName, apiToken, domain string, maxRetries uint64) *BulkImporter {
+	return &BulkImporter{
+		//url:        fmt.Sprintf("https://%s.zendesk.com/api/v2/imports/tickets/create_many", domain),
+		client:     newHTTPClient(),
+		userName:   userName,
+		apiToken:   apiToken,
+		baseURL:    fmt.Sprintf("https://%s.zendesk.com", domain),
+		maxRetries: maxRetries,
+	}
+}
+
+// Write buffer data to zendesk
+func (b *BulkImporter) Write(ctx context.Context, records []sdk.Record) error {
+	bufferedTicket, err := parseRecords(records)
+	if err != nil {
+		return fmt.Errorf("unable to parse the records %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(
+		ctx,
+		http.MethodPost,
+		fmt.Sprintf("%s/api/v2/imports/tickets/create_many", b.baseURL),
+		bytes.NewBuffer(bufferedTicket),
+	)
+
+	if err != nil {
+		return fmt.Errorf("unable to send to zendesk server %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json; charset=UTF-8")
+	req.Header.Add("Authorization", "Basic "+basicAuth(b.userName, b.apiToken))
+
+	resp, err := b.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("unable to fetch response from zendesk server %w", err)
+	}
+
+	defer resp.Body.Close()
+	// Validation for httpStatusCode 429 - Too many Requests, Retry value after `93s`
+	if resp.StatusCode == http.StatusTooManyRequests {
+		// NOTE: https://developer.zendesk.com/documentation/ticketing/using-the-zendesk-api/best-practices-for-avoiding-rate-limiting/#catching-errors-caused-by-rate-limiting
+		retryValue, err := strconv.ParseInt(resp.Header.Get("Retry-After"), 10, 64)
+		if err != nil {
+			return fmt.Errorf("unable to get retry value: %w", err)
+		}
+
+		sdk.Logger(ctx).Trace().Int64("Retry-After", retryValue).Msg("rate limit exceeded, will retry after `Retry-After` duration")
+
+		if b.retryCount >= b.maxRetries {
+			return fmt.Errorf("rate-limit exceeded, total retries: %d", b.retryCount)
+		}
+
+		b.retryCount++
+		// retry writing after the cool off duration passes, block till then
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(time.Duration(retryValue) * time.Second):
+			return b.Write(ctx, records)
+		}
+	}
+
+	// reset the retry count, in case of non 429 response.
+	b.retryCount = 0
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("non 200 status code received(%v)", resp.StatusCode)
```

Comment:
Returning the response line and/or response body in case of error is very useful too.